src/policy/v3_policy.cpp

```diff
@@ -0,0 +1,197 @@
+
+    // Now we have all ancestors, so we can start checking v3 rules.
+    if (ptx->nVersion == 3) {
+        if (mempool_ancestors.size() + in_package_parents.size() + 1 > V3_ANCESTOR_LIMIT) {
+            return strprintf("tx %s would have too many ancestors", ptx->GetWitnessHash().ToString());
+        }
+
+        const bool has_parent{mempool_ancestors.size() + in_package_parents.size() > 0};
+        if (has_parent) {
+            // A v3 child cannot be too large.
+            if (vsize > V3_CHILD_MAX_VSIZE) {
+                return strprintf("v3 child tx %s is too big: %u > %u virtual bytes",
+                                 ptx->GetWitnessHash().ToString(), vsize, V3_CHILD_MAX_VSIZE);
+            }
+
+            // Find the parent and extract the information we need for v3 checks.
+            int parent_version = 0;
+            // Used to match parents by prevout
+            Txid parent_hash = Txid::FromUint256(uint256(0));
+            // Used for the error string.
+            Wtxid parent_wtxid = Wtxid::FromUint256(uint256(0));
+            bool has_mempool_sibling{false};
+
+            if (mempool_ancestors.size() > 0) {
+                // There's a parent in the mempool.
+                auto& parent = *mempool_ancestors.begin();
+                parent_version = parent->GetTx().nVersion;
+                has_mempool_sibling = parent->GetCountWithDescendants() > 1;
+                parent_hash = parent->GetTx().GetHash();
+                parent_wtxid = parent->GetTx().GetWitnessHash();
+            } else {
+                // Ancestor must be in the package. Find it.
+                auto &parent_index = in_package_parents[0];
+                // If the in-package parent has mempool ancestors, then this is
+                // a v3 violation.
+                if (package_with_ancestors.ancestor_counts[parent_index] > 0) {
+                    return strprintf("tx %s would have too many ancestors", ptx->GetWitnessHash().ToString());
+                }
+
+                auto &parent = package_with_ancestors.package[parent_index];
+                parent_version = parent->nVersion;
+                parent_hash = parent->GetHash();
+                parent_wtxid = parent->GetWitnessHash();
+            }
+
+            // If there is a parent, it must have the right version.
+            if (parent_version != 3) {
+                return strprintf("v3 tx %s cannot spend from non-v3 tx %s",
+                                 ptx->GetWitnessHash().ToString(), parent_wtxid.ToString());
+            }
+
+            // The mempool or in-package parent cannot have any other in-mempool children. We don't
+            // check whether the sibling is to-be-replaced (done in ApplyV3Rules) because that
+            // doesn't apply in a package.
+            if (has_mempool_sibling) {
+                return strprintf("tx %u would exceed descendant count limit", parent_wtxid.ToString());
```

Comment:
Removed the duplicate check, nice

Reply:
double-checking: are we sure we can't hit this? @sdaftuar

Reply:
I think indeed we cannot hit this, but it's because we call `ApplyV3Rules` in a package.

When `package` is a parent+child, `has_mempool_sibling` doesn't matter since having any mempool ancestor should cause this to fail immediately. In that way, this function kind of assumes that `package` is a connected component.

If `package` has unrelated transactions, we'd let a mempool sibling slip through in this function. However, since we are calling `ApplyV3Rules` within `PreChecks`, the mempool sibling is caught there. So I guess the "not strictly necessary in package validation" is not really correct for our "testmempoolaccept is allowed to have unrelated transactions" use case.

Reply:
@glozow's analysis looks right to me. I don't think it matters too much way we go -- either we should update the code comments to explain that correctness relies on invoking ApplyV3Rules and PV3C on package transactions, or we can just leave the duplicate check in here so that PV3C stands on its own for package validation.  I'd probably vote for just updating the comment but I don't feel strongly either way.

Reply:
Leaving the check in with an `Assume` if we don't expect to hit it probably best, with a comment?

Reply:
Put the check in with an `Assume` and comment. Also renamed `ApplyV3Rules` to `SingleV3Checks`